device-connectors/src/testflinger_device_connectors/devices/__init__.py

```diff
@@ -61,6 +61,20 @@ class RecoveryError(Exception):
     pass
 
 
+def log_provision_error(exception: Exception):
+    exception_info = {
+        "exception_info": {
+            "exception_name": type(exception).__name__,
+            "exception_message": str(exception),
+            "exception_cause": repr(exception.__cause__),
```

Comment:
Just out of curiosity, what kinds of things have you managed to catch in the __cause__ field when you tried this with exceptions from maas? From what I understand, in other circumstances (chaining exceptions) it might be more beneficial to check `__context__` if `__cause__` isn't present. Also, I wonder if `__doc__` might actually prove more useful? It *should* normally be more human readable, but might lack details that cause or context might have if they exist at all.

Reply:
I used cause here in order to capture exceptions like MaasStorageError which seemed to be raised as `raise ProvisioningError from MaasStorageError`

I'll take a look at doc and context.